### PR TITLE
Skip updating user and group ID for appsubs

### PR DESCRIPTION
**Pull request by philipwu08**
https://github.com/open-cluster-management/backlog/issues/9533

Signed-off-by: Philip Wu <phwu@redhat.com>

### CODE REVIEW

**Comment by jnpacker** [comment]
/lgtm (i'm not in owners)

* @qiujian16 are you okay with the const, or have an alternative suggestion?
* testcase if there are already FVT available

File: pkg/webhook/useridentity/mutatingWebhook.go [hunk 1]
```diff
@@ -92,6 +95,17 @@ func (a *AdmissionHandler) mutateResource(ar v1.AdmissionReview) *v1.AdmissionRe
 	}
 
 	annotations := obj.GetAnnotations()
+
+	// Do not change the userInfo if the object is being created by the AppMgr.
+	// This value is stamped onto the Subscription object by Appmgr (Only for subscriptions)
+	klog.V(2).Infof("Object kind:%s, User request:%s, expected:%s", obj.GetKind(), ar.Request.UserInfo.Username, UserIDforAppMgr)
+	if obj.GetKind() == "Subscription" && ar.Request.UserInfo.Username == UserIDforAppMgr {
```
Comment (qiujian16) [inline pkg/webhook/useridentity/mutatingWebhook.go:102]: I would rather have a flag to specify which identity should be skipped, and has UserIDforAppMgr as the default value.

Comment (philipwu08) [reply pkg/webhook/useridentity/mutatingWebhook.go:102]: Quijing, thanks for the review. This is all new to me. Can you give me more details about this flag? Are you asking this flag to be passed in the AdmissionRequest?

Comment (qiujian16) [reply pkg/webhook/useridentity/mutatingWebhook.go:102]: https://github.com/open-cluster-management/multicloud-operators-foundation/blob/main/cmd/webhook/app/options/options.go#L33 add a flag entry here

Comment (philipwu08) [reply pkg/webhook/useridentity/mutatingWebhook.go]: Quijian, thank you for the hint. Hopefully I got it right. Please review. Thanks.

File: cmd/webhook/app/options/options.go [hunk 2]
```diff
@@ -12,21 +12,23 @@ import (
...
-		CertFile:       "",
-		KeyFile:        "",
-		QPS:            100.0,
-		Burst:          200,
+		KubeConfigFile:    "",
+		CertFile:          "",
+		KeyFile:           "",
+		QPS:               100.0,
+		Burst:             200,
+		UserSkipOverwrite: "system:serviceaccount:open-cluster-management-agent-addon:klusterlet-addon-appmgr",
```
Comment (jnpacker) [inline cmd/webhook/app/options/options.go:31]: Since we have this open, did you want to make it support a list of users?

Comment (philipwu08) [reply cmd/webhook/app/options/options.go:31]: It crossed my mind as well, not sure if we have a use-case for it. I'll work on having the changes ready, but not push the commit until Qiujian agrees.

Comment (qiujian16) [reply cmd/webhook/app/options/options.go:31]: lets make it a list

File: pkg/webhook/useridentity/mutatingWebhook.go [hunk 3]
```diff
@@ -87,6 +86,7 @@ func (a *AdmissionHandler) serve(w io.Writer, r *http.Request, admit admitFunc)
 
 func (a *AdmissionHandler) mutateResource(ar v1.AdmissionReview) *v1.AdmissionResponse {
 	klog.V(2).Info("mutating custom resource")
```
Comment (qiujian16) [inline pkg/webhook/useridentity/mutatingWebhook.go:88]: We always use v(4) here

File: pkg/webhook/useridentity/mutatingWebhook.go [hunk 4]
```diff
@@ -92,6 +95,14 @@ func (a *AdmissionHandler) mutateResource(ar v1.AdmissionReview) *v1.AdmissionRe
 	}
 
 	annotations := obj.GetAnnotations()
+
+	if ar.Request.UserInfo.Username == a.UserSkipOverwrite {
+		klog.V(2).Infof("Skip add user and group for resource: %+v, name: %+v", ar.Request.Resource.Resource, obj.GetName())
```
Comment (qiujian16) [inline pkg/webhook/useridentity/mutatingWebhook.go:100]: v(4) here also

File: pkg/webhook/useridentity/mutatingWebhook.go (hunk 3 above)
Comment (philipwu08) [reply pkg/webhook/useridentity/mutatingWebhook.go]: Thanks Quijian. New commit has the requested changes. Not really sure where to put the "contains" functions - it probably belongs in the util package but couldn't find a good home for it.

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm
/hold
 
hold in case you want others to review it, otherwise feel free to cancel it. BTW, it is better to squash before merge it.

File: pkg/webhook/useridentity/mutatingWebhook.go (hunk 3 above)
Comment (qiujian16) [reply pkg/webhook/useridentity/mutatingWebhook.go]: there is a ContainsString in util pkg.

**Comment by qiujian16** [comment]
thanks, would you like to squash the commit?

**Comment by philipwu08** [comment]
squashed -thanks Quijian!

**Pull request merged**